examples/docs_projects/project_components_pdf_extraction/pdf_extraction/lib/pdf_extraction.py

```diff
@@ -0,0 +1,273 @@
+        openai_api_key (str): OpenAI API key for validation
+        openai_model (str): OpenAI model to use for validation
+        validation_score (int): Minimum acceptable validation score (1-10)
+    """
+
+    pdf_dir: str
+    output_dir: str
+    validation_score: int = 7
+    language: str = "eng"
+    dpi: int = 300
+    openai_model: str = "gpt-4o-mini"
+
+    def _normalize_key(self, key: str) -> str:
+        return key.replace("-", "_")
+
+    def build_defs(self, context: ComponentLoadContext) -> dg.Definitions:
+        # Get all PDF files from the directory
+        if os.path.isdir(self.pdf_dir):
+            pdf_files = [f for f in os.listdir(self.pdf_dir) if f.lower().endswith(".pdf")]
+            pdf_paths = [os.path.join(self.pdf_dir, pdf) for pdf in pdf_files]
+        else:
+            # If a single file is provided, wrap it in a list
+            pdf_paths = [self.pdf_dir]
+
+        # Create single shared PDF extractor resource
+        pdf_extractor_resource = PDFTextExtractor(
+            language=self.language,
+            dpi=self.dpi,
+            openai_api_key=dg.EnvVar("OPENAI_API_KEY"),
+            preprocess=True,
+            output_dir=self.output_dir,  # Base output directory
+        )
+
+        assets = []
+        asset_checks = []
+        all_jobs = []
+
+        # Create assets for each PDF file
+        for pdf_path in pdf_paths:
+            key_prefix = self._normalize_key(f"{Path(pdf_path).stem}")
+
+            # Create convert_to_image asset with captured pdf_path
+            def _make_convert_to_image(pdf_path=pdf_path, key_prefix=key_prefix):
+                @dg.asset(name=f"{key_prefix}_convert_to_image")
+                def convert_to_image(
+                    context: dg.AssetExecutionContext, pdf_extractor: PDFTextExtractor
+                ):
+                    """Convert PDF to images, one per page."""
+                    return pdf_extractor.convert_pdf_to_images(
+                        pdf_path, output_folder=os.path.join(self.output_dir, key_prefix)
+                    )
+
+                return convert_to_image
+
+            # Create extract_text asset with captured key_prefix
+            def _make_extract_text(key_prefix=key_prefix, pdf_path=pdf_path):
+                @dg.asset(
+                    name=f"{key_prefix}_extract_text",
+                    deps=[f"{key_prefix}_convert_to_image"],
+                )
+                def extract_text(
+                    context: dg.AssetExecutionContext, pdf_extractor: PDFTextExtractor
+                ):
+                    """Extract text from the converted images using OCR."""
+                    context.log.info(f"Extracting text for PDF: {key_prefix}")
+
+                    # Define the output directory and ensure it exists
+                    pdf_output_dir = os.path.join(self.output_dir, key_prefix)
+                    os.makedirs(pdf_output_dir, exist_ok=True)
```

Comment:
Does this check need to happen? Are there cases where the upstream asset will not create the folder?